core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -2570,10 +2570,20 @@ class SparkContext(config: SparkConf) extends Logging {
 
   private[spark] def newShuffleId(): Int = nextShuffleId.getAndIncrement()
 
-  private val nextRddId = new AtomicInteger(0)
+  private var nextRddId = new AtomicInteger(0)
 
   /** Register a new RDD, returning its RDD ID */
-  private[spark] def newRddId(): Int = nextRddId.getAndIncrement()
+  private[spark] def newRddId(): Int = {
+    var id = nextRddId.getAndIncrement()
+    if (id >= 0) {
+        return id
+    }
+    this.synchronized {
+      nextRddId = new AtomicInteger(0)
+      id = nextRddId.getAndIncrement()
+    }
+    id
+  }
```

Comment:
Something like this would remove the need for synchronization, etc

```suggestion
  private[spark] def newRddId(): Int = {
    nextRddId.getAndUpdate { i =>
      var nextValue = i + 1
      if (nextValue < 0) {
        nextValue = 0
      }
      nextValue
    }
  }
```